onnxruntime/python/tools/quantization/matmul_4bits_quantizer.py

```diff
@@ -165,20 +264,102 @@ def _process_subgraph(self, graph_stack: List[GraphProto]):
         graph_stack.pop()
         return graph
 
+    def _generate_q4_node_config(self):
+        """Generate weight only quant configuration for nodes."""
+        q4_node_config = {}
+        template_config_q4 = {"bits": 4, "group_size": self.block_size, "scheme": "sym" if self.is_symmetric else "asym"}
+        template_config_fp32 = "fp32"
+        for node in self.model.model.graph.node:
+            if node.op_type in ["MatMul"]:
+                if not all([self.model.get_initializer(i) is None for i in node.input]):
+                    q4_node_config[node.name] = template_config_q4
+                else:
+                    q4_node_config[node.name] = template_config_fp32
+        return q4_node_config
+
+    def int4_quant_algo(self):
+        """4b quantize a model with RTN or GPTQ algorithm. Please refer to
+        https://github.com/intel/neural-compressor/blob/master/docs/source/quantization_weight_only.md
+        for more details on weight only quantization using Intel® Neural Compressor.
+        """
+
+        def inc_dataloader():
+            data_reader = copy.deepcopy(self.algo_config.calibration_data_reader)
+            for data in data_reader:
+                yield data, None
+
+        accuracy_level = self.algo_config.accuracy_level
+        weight_only_node_config = self._generate_q4_node_config()
+
+        algorithm = self.algo_config.algorithm
+        logger.info(f"start to quantize model with {algorithm} algorithm...")
+        if algorithm == "RTN":
+            from neural_compressor.adaptor.ox_utils.weight_only import rtn_quantize
+
+            ratios = self.algo_config.ratios
+
+            self.model = rtn_quantize(
+                model=self.model_path if self.model_path is not None else self.model.model,
+                weight_config=weight_only_node_config,
+                ratios=ratios,
+                accuracy_level=accuracy_level,
+            )
+        elif algorithm == "GPTQ":
+            from neural_compressor.adaptor.ox_utils.weight_only import gptq_quantize
+
+            percdamp = self.algo_config.percdamp
+            blocksize = self.algo_config.blocksize
+            actorder = self.algo_config.actorder
+            mse = self.algo_config.mse
+            perchannel = self.algo_config.perchannel
+            dataloader = inc_dataloader()
+
+            self.model = gptq_quantize(
```

Comment:
>gptq_quantize [](http://example.com/codeflow?start=25&length=13)

does intel-compressor use all available providers? It'd better to do so. In this way, it will leverage OpenVino and other providers:
[https://github.com/microsoft/onnxruntime/blob/98510fb8fb2d761aad73ed8d32d4a6922546bdf0/docs/python/tutorial.rst#step-3-load-and-run-the-model-using-onnx-runtime](https://github.com/microsoft/onnxruntime/blob/98510fb8fb2d761aad73ed8d32d4a6922546bdf0/docs/python/tutorial.rst#step-3-load-and-run-the-model-using-onnx-runtime)

Reply:
I tried the example for llama. It is quite slow to quantize.

Reply:
Inten Neural Compressor does provide the popular quantization algorithms, while these algorithms are usually device/provider agnostic. As long as the quantized model is compliant, it is supposed to run well with other provides.

Reply:
How to enable other providers or devices?

Reply:
Hi yufeng, it depends on the kernel readiness on other providers

Reply:
> I tried the example for llama. It is quite slow to quantize.

gptq quantize need to dump the input of matmul, and we do it one by one to avoid OOM https://github.com/intel/neural-compressor/blob/4e88ba48e889494c80cfe566243b89e839c528a9/neural_compressor/adaptor/ox_utils/weight_only.py#L1060C53-L1060C53